### PR TITLE
Migrate to `fvm_shared` lib

### LINKED ISSUE #546: Unable to build lib for windows
clone and run `cargo build` in windows, many errors appeared

```
warning: cl : Command line warning D9024 : unrecognized source file type './asm/mul_4.S', object file assumed
warning: cl : Command line warning D9027 : source file './asm/mul_4.S' ignored

error: failed to run custom build command for `fff v0.2.3`

Caused by:
  process didn't exit successfully: `D:\repos\brave\brave-browser\src\out\Component\brave_rust/out/x64\debug\build\fff-ad279d28292ce510\build-script-build` (exit code: 1)
  --- stdout
  TARGET = Some("x86_64-pc-windows-msvc")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-pc-windows-msvc")
  CC_x86_64-pc-windows-msvc = None
  CC_x86_64_pc_windows_msvc = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-pc-windows-msvc = None
  CFLAGS_x86_64_pc_windows_msvc = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,sse,sse2")
  DEBUG = Some("true")
  running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-W4" "-c" "-FoD:\\repos\\brave\\brave-browser\\src\\out\\Component\\brave_rust/out/x64\\x86_64-pc-windows-msvc\\debug\\build\\fff-139bd3b46806c1c7\\out\\./asm/mul_4.o" "-c" "./asm/mul_4.S"
  cargo:warning=cl : Command line warning D9024 : unrecognized source file type './asm/mul_4.S', object file assumed
  cargo:warning=cl : Command line warning D9027 : source file './asm/mul_4.S' ignored
  cl : Command line warning D9021 : no action performed
  exit code: 0
  AR_x86_64-pc-windows-msvc = None
  AR_x86_64_pc_windows_msvc = None
  HOST_AR = None
  AR = None
  running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\lib.exe" "-out:D:\\repos\\brave\\brave-browser\\src\\out\\Component\\brave_rust/out/x64\\x86_64-pc-windows-msvc\\debug\\build\\fff-139bd3b46806c1c7\\out\\libff-derive-crypto.a" "-nologo" "D:\\repos\\brave\\brave-browser\\src\\out\\Component\\brave_rust/out/x64\\x86_64-pc-windows-msvc\\debug\\build\\fff-139bd3b46806c1c7\\out\\./asm/mul_4.o"
  LINK : fatal error LNK1181: cannot open input file 'D:\repos\brave\brave-browser\src\out\Component\brave_rust\out\x64\x86_64-pc-windows-msvc\debug\build\fff-139bd3b46806c1c7\out\.\asm\mul_4.o'
  exit code: 1181

  --- stderr


  error occurred: Command "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX64\\x64\\lib.exe" "-out:D:\\repos\\brave\\brave-browser\\src\\out\\Component\\brave_rust/out/x64\\x86_64-pc-windows-msvc\\debug\\build\\fff-139bd3b46806c1c7\\out\\libff-derive-crypto.a" "-nologo" "D:\\repos\\brave\\brave-browser\\src\\out\\Component\\brave_rust/out/x64\\x86_64-pc-windows-msvc\\debug\\build\\fff-139bd3b46806c1c7\\out\\./asm/mul_4.o" with args "lib.exe" did not execute successfully (status code exit code: 1181).
```
afaik it goes from https://github.com/ChainSafe/forest dependency

### PR DESCRIPTION
**Pull request by rllola**
- We are dropping Chainsafe lib because there is no update anymore.
- Fix #546 
- JSON serialization for parameters and message now supported (see `extras` folder)
- Remove pure js `transactionSignLotus` as we now only support Message using Lotus CamelCase standard

Notes: There is a lot of changes and some breaking ones. Now the json format only support PascalCase like it is standard in Lotus.

### CODE REVIEW

File: examples/wasm_node/test/tests_basic.js [hunk 1]
```diff
@@ -229,6 +229,7 @@ describeCall('keyRecoverBLS', function() {
 
 describe('transactionSerialize', function() {
   it('should serialize transaction', function() {
+    console.log(dataTxs[0].transaction)
```
Comment (Anonyma) [inline examples/wasm_node/test/tests_basic.js:232]: `console.log`

File: examples/wasm_node/test/tests_basic.js [hunk 2]
```diff
@@ -274,7 +275,7 @@ describe('transactionParse', function() {
 
     assert.throws(
       () => filecoin_signer.transactionParse(cbor_transaction_extra_bytes, false),
-      /Error: CBOR decode error: too many terminals, data makes no sense|Encoding error \| trailing data at offset 64/,
+      /Cannot deserilaize parameters \| Serialization error for Cbor protocol: trailing data at offset 64|CBOR decode error: too many terminals, data makes no sense/,
```
Comment (Anonyma) [inline examples/wasm_node/test/tests_basic.js:278]: typo `deserilaize`

File: examples/wasm_node/test/tests_basic.js [hunk 3]
```diff
@@ -284,7 +285,7 @@ describe('transactionParse', function() {
...
```
Comment (Anonyma) [inline examples/wasm_node/test/tests_basic.js:288]: (same typo)

File: hfuzz-signer/signed-message.rs [hunk 4]
```diff
@@ -1,20 +1,5 @@
...
-struct DummySigner;
-
-impl Signer for DummySigner {
-    fn sign_bytes(&self, _: &[u8], _: &Address) -> Result<Signature, Box<dyn std::error::Error>> {
-        Ok(Signature::new_secp256k1([0u8].to_vec()))
-    }
-}
-
 fn main() {
+    /* FIXME
```
Comment (becominginsane) [inline hfuzz-signer/signed-message.rs:2]: what's the issue encountered here?

File: hfuzz-signer/unsigned-message.rs [hunk 5]
```diff
@@ -1,8 +1,5 @@
-use core::convert::TryFrom;
-use filecoin_signer::api::{MessageTx, MessageTxAPI, MessageTxNetwork, UnsignedMessageAPI};
-use forest_message::UnsignedMessage;
-
 fn main() {
+    /* FIXME
```
Comment (becominginsane) [inline hfuzz-signer/unsigned-message.rs:2]: maybe we should reinstate these if it's applicable so we don't lose fuzzing coverage

File: signer-npm/src/lib.rs [hunk 6]
```diff
@@ -185,14 +183,21 @@ pub fn transaction_serialize_raw(unsigned_message: JsValue) -> Result<Vec<u8>, J
     set_panic_hook();
 
     // TODO: Should be MessageTxAPI because it can be unsigned message or signed message
```
Comment (becominginsane) [inline signer-npm/src/lib.rs:185]: this TODO can be removed now :smiley_cat: :clap:

File: signer-npm/src/lib.rs [hunk 7]
```diff
@@ -223,13 +228,19 @@ pub fn transaction_sign(
         .into_serde()
         .map_err(|e| JsValue::from(format!("Error parsing parameters: {}", e)))?;
 
+    let msg = match unsigned_message {
+        MessageTxAPI::Message(m) => m,
+        MessageTxAPI::SignedMessage(_) => {
+            return Err(JsValue::from("Can't serialize SignedMessage"))
```
Comment (becominginsane) [inline signer-npm/src/lib.rs:234]: can this error message be made clearer? the function is called "transaction_sign" so this error doesn't seem to bring clarity of why the function errored, perhaps "Attempted to sign a SignedMessage" would be more explicit

File: signer-npm/src/lib.rs [hunk 8]
```diff
@@ -246,13 +257,22 @@ pub fn transaction_sign_lotus(
...
```
Comment (becominginsane) [inline signer-npm/src/lib.rs:263]: same as earlier, "Attempted to sign a SignedMessage" would make the error clearer

File: signer-npm/src/lib.rs [hunk 9]
```diff
@@ -246,13 +257,22 @@ pub fn transaction_sign_lotus(
...
-    let signed_message =
-        filecoin_signer::transaction_sign(&unsigned_message, &private_key_bytes)
-            .map_err(|e| JsValue::from_str(format!("Error signing transaction: {}", e).as_str()))?;
+    let signed_message = filecoin_signer::transaction_sign(&msg, &private_key_bytes)
+        .map_err(|e| JsValue::from_str(format!("Error signing transaction: {}", e).as_str()))?;
 
-    let signed_message_lotus = utils::convert_to_lotus_signed_message(signed_message)?;
+    let signed_message_lotus = serde_json::to_string(&MessageTxAPI::SignedMessage(signed_message))
+        .map_err(|e| {
+            JsValue::from_str(format!("Error converting the into JSON: {}", e).as_str())
```
Comment (becominginsane) [inline signer-npm/src/lib.rs:274]: Error converting the ... ? into JSON

File: signer-npm/src/lib.rs [hunk 10]
```diff
@@ -268,15 +288,22 @@ pub fn transaction_sign_raw(
         .into_serde()
         .map_err(|e| JsValue::from(format!("Error parsing parameters: {}", e)))?;
 
+    let msg = match unsigned_message {
+        MessageTxAPI::Message(m) => m,
+        MessageTxAPI::SignedMessage(_) => {
+            return Err(JsValue::from("Can't serialize SignedMessage"))
```
Comment (becominginsane) [inline signer-npm/src/lib.rs:294]: Attempted to sign a SignedMessage

File: signer/src/api.rs [hunk 11]
```diff
@@ -117,67 +80,32 @@ impl MessageParams {
...
 }
 
-/// Signed message api structure
-#[cfg_attr(feature = "with-arbitrary", derive(arbitrary::Arbitrary))]
-#[derive(Debug, Clone, Deserialize, PartialEq, Serialize)]
-pub struct SignedMessageAPI {
-    pub message: UnsignedMessageAPI,
-    pub signature: SignatureAPI,
-}
+impl TryFrom<MessageTx> for MessageTxAPI {
```
Comment (becominginsane) [inline signer/src/api.rs:100]: Is this even fallible? It doesn't look like it is so why not `impl From<MessageTx> for MessageTxAPI` ?

**Review by becominginsane — APPROVED** [review]
LGTM :clap: A lot of work, glad to also see that you used clippy

File: hfuzz-signer/signed-message.rs (hunk 4 above)
Comment (rllola) [reply hfuzz-signer/signed-message.rs:2]: The fuzzer was on a now removed structure. We now use `SignedMessage` from `extras`. I didn't want to include this in the PR as it is already big and I haven't done much fuzzing work.

I will create a issue for re-introduce fuzzing. Also we now used directly structure from the fvm_shared lib so they might be already doing just that.

File: hfuzz-signer/unsigned-message.rs (hunk 5 above)
Comment (rllola) [reply hfuzz-signer/unsigned-message.rs:2]: I will create an issue to have this cover. And as mentioned we are now reusing directly the fvm_shared structures.

**Pull request merged**